aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -1054,8 +1054,17 @@ static Tensor maybe_add_maybe_sub(const Tensor& self, const Tensor& other, const
   // hack to use the TensorIterator to get the correct broadcasting and type promotion logic
   auto device_ = Device(DeviceType::Meta);
   constexpr c10::DispatchKeySet meta_dks(at::DispatchKey::Meta);
+  // We can't just call self.to(device_) here. Why?
+  // We **need** to make sure that the CompositeImplicitAutograd to() kernel runs, so that
+  // if there are any subclasses, they always see the desguared to() call.
+  // We are in a ZeroTensor kernel below autograd, though, which means that `.to()`
+  // will hit the Python key before hitting the composite kernel.
```

Comment:
So I hit this when I first wrote some basic manual tests for zero tensor (I *think* my test in `test_functionalization.py` actually exercises this, but let me double check and add a better test if not).

The problem is that this function is registered to the `ZeroTensor` dispatch key. The relevant dispatch keys are something like:
```
AutogradCPU (autograd has already run)
ZeroTensor <-- we are here
Python <-- we will dispatch here next
CPU
```

`aten::to()` has a `CompositeImplicitAutograd` registration. `CompositeImplicitAutograd` is an alias key that effectively registers the function to all autograd keys, and all "backend" keys (*not* including python, since python isn't a backend key).

So when we call `aten::to()` here, the next key we hit is the Python key, and we never get a chance to hit a key that runs the CompositeImplicitAutograd kernel.

Reply:
Hmm, maybe a good question is why don't we just add the `Python` key to the `CompositeImplicitAutograd` alias keyset today? I'm not actually sure.

I'd be happy to try making that change. Although I'm a bit worried that it might be higher blast radius. If it's ok with you, I'm accumulating a list of followups that I want to revisit once I've unblocked DTensor compile support, and I can add this to the list. What do you think?

Reply:
Does C++ functionalization have the same behavior?

Reply:
Hmm, we don't actually want to disable the python key (why? We still want FakeTensor to run as part of computing `meta_out` below, since we need it e.g. in case the output has symbolic shapes).

I think in theory we'd want something like a C++ variant of `disable_proxy_modes_tracing()` (to just turn off tracing but nothing else).

I'll file this under the list of things that we should fix, but aren't needed to land python functionalization though. So if no tests are passing, and I get the same trace out with C++ vs python functionalization, I'll leave it for now.

Reply:
It looks like this is because DispatchKey::Functionalize is a part of the CompositeImplicitAutograd keyset

Reply:
sync'd offline - totally agreed, will change this (although in the near future when we turn on functionalization for pre_dispatch, we'll want a way to let functionalization not run these decomps)